src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java

```diff
@@ -81,4 +81,142 @@ public SearchResultPageDTO<LearningGoal> getAllOnPageWithSize(final PageableSear
+                sourceVertex.setBeingVisited(true);
+                for (Vertex neighbor : sourceVertex.getAdjacencyList()) {
+                    if (neighbor.isBeingVisited()) {
+                        // backward edge exists
+                        return true;
+                    }
+                    else if (!neighbor.isVisited() && vertexIsPartOfCycle(neighbor)) {
+                        return true;
+                    }
+                }
+                sourceVertex.setBeingVisited(false);
+                sourceVertex.setVisited(true);
+                return false;
+            }
+        }
+
+        var graph = new Graph();
+        for (LearningGoal learningGoal : learningGoals) {
+            graph.addVertex(new Vertex(learningGoal.getTitle()));
+        }
+        for (LearningGoalRelation relation : relations) {
+            var headVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getHeadLearningGoal().getTitle())).findFirst().orElseThrow();
+            var tailVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getTailLearningGoal().getTitle())).findFirst().orElseThrow();
+            // Only EXTENDS and ASSUMES are included in the generated graph as other relations are no problem if they are circular
+            // MATCHES relations are considered in the next step by merging the edges and combining the adjacencyLists
+            switch (relation.getType()) {
+                case EXTENDS, ASSUMES -> graph.addEdge(tailVertex, headVertex);
+            }
+        }
+        // combine vertices that are connected through MATCHES
+        for (LearningGoalRelation relation : relations) {
+            if (relation.getType() == LearningGoalRelation.RelationType.MATCHES) {
+                var headVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getHeadLearningGoal().getTitle())).findFirst().orElseThrow();
+                var tailVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getTailLearningGoal().getTitle())).findFirst().orElseThrow();
+                if (headVertex.adjacencyList.contains(tailVertex) || tailVertex.adjacencyList.contains(headVertex)) {
+                    return true;
+                }
+                // create a merged vertex
+                var mergedVertex = new Vertex(tailVertex.label + ", " + headVertex.label);
+                // add all neighbours to merged vertex
+                for (Vertex vertex : headVertex.getAdjacencyList()) {
+                    mergedVertex.addNeighbor(vertex);
+                }
+                for (Vertex vertex : tailVertex.getAdjacencyList()) {
+                    mergedVertex.addNeighbor(vertex);
+                }
+                // update every vertex that initially had one of the two merged vertices as neighbours to now reference the merged vertex
+                for (Vertex vertex : graph.vertices) {
```

Comment:
Wouldn't it be enough to iterate over the neighbors of the two vertices that were merged?